hw_diag/utilities/network_watchdog.py

```diff
@@ -0,0 +1,169 @@
+        try:
+            socket.setdefaulttimeout(timeout)
+            sock = socket.create_connection((self.PUBLIC_DNS_SERVER, self.PUBLIC_DNS_PORT))
+            sock.close()
+            return True
+        except Exception as e:
+            self.LOGGER.info(f"internet not accessible: {e}")
+            return False
+
+    def is_connected(self) -> bool:
+        self.LOGGER.info("Checking the network connectivity.")
+
+        # Log more details about the network connectivity and internet connectivity
+        self.LOGGER.info(f"Network manager state: {self.network_manager.get_connect_state()}")
+        self.LOGGER.info(f"Internet connectivity: {self.have_internet()}")
+
+        nm_connected = self.network_manager.is_connected()
+        return nm_connected
+
+    def restart_network_manager(self):
+        """Restart hostOS NetworkManager service"""
+        nm_restarted = self.network_manager_unit.wait_restart()
+        self.LOGGER.info(f"Network manager restarted: {nm_restarted}")
+
+    def get_last_restart(self) -> datetime:
+        try:
+            last_restart = KeyStore(self.state_file_path).get(self.LAST_RESTART_KEY)
+            return datetime.strptime(last_restart, self.LAST_RESTART_DATE_FORMAT)
+        except Exception as e:
+            self.LOGGER.info(f"Can not find the previous restart time: {e}")
+            return datetime.min
+
+    def save_last_restart(self) -> None:
+        store = KeyStore(self.state_file_path)
+        store.set(self.LAST_RESTART_KEY, datetime.now().strftime(self.LAST_RESTART_DATE_FORMAT))
+        self.LOGGER.info("Saved the last_restart.")
+
+    def ensure_network_connection(self) -> None:
+        self.LOGGER.info("Running the watchdog...")
+
+        # If network is connected, nothing to do more
+        if self.is_connected():
+            self.lost_count = 0
+            self.LOGGER.info("Network is working.")
+            return
+
+        # If network is not working, take the next step
+        self.lost_count += 1
+        self.LOGGER.warning(
+            f"Network is not connected! Lost connectivity count={self.lost_count}")
+
+        if self.lost_count > self.NM_RESTART_THRESHOLD:
+            self.LOGGER.warning(
+                "Reached threshold for nm restart for recovering network.")
+            self.restart_network_manager()
+            self.LOGGER.info("Restarted the network connection.")
+
+        if self.lost_count > self.FULL_REBOOT_THRESHOLD:
+            self.LOGGER.warning(
+                "Reached threshold for system reboot for recovering network.")
+            if datetime.now() - timedelta(
+                    hours=self.REBOOT_LIMIT_HOURS) < self.get_last_restart():
+                self.LOGGER.info(
+                    "Hotspot has already been restarted within a day, skipping.")
+                return
+
+            self.save_last_restart()
+
+            force_reboot = self.lost_count > self.FULL_FORCE_REBOOT_THRESHOLD
```

Comment:
I think the logic for forcing restart  (after regular restart fails) will not work. save_last_restart() will make sure that we never reach force_reboot = True

Reply:
Ok, up time looks better than saved last_restart time.